staging/src/k8s.io/dynamic-resource-allocation/structured/internal/stable/pools_stable.go

```diff
@@ -139,8 +97,33 @@ func GatherPools(ctx context.Context, slices []*resourceapi.ResourceSlice, node
 
 	// Find incomplete pools and flatten into a single slice.
 	result := make([]*Pool, 0, len(pools))
-	for _, pool := range pools {
-		pool.IsIncomplete = incompletePools.Has(poolIdentifier{driver: pool.Driver.String(), pool: pool.Pool.String()})
+	for poolID, pool := range pools {
+		isIncomplete := int64(len(pool.Slices)) != pool.Slices[0].Spec.Pool.ResourceSliceCount
+		// If we have all slices, we are done. If not, then we need to start looking for slices
+		// which were filtered out above because their node selection made them look irrelevant
+		// for the current node. This is necessary for "allocate all" mode (it rejects incomplete
+		// pools).
+		if isIncomplete {
+			// Find all the slices that belongs to the current pool. Doing this for each pool
+			// separately could be a big inefficient if there are multiple incomplete pools. But
+			// that is probably a rare situation.
+			slicesForPool := findSlicesForPool(slices, poolID)
+			// This allows to check ALL slices in the pool to make sure whether the pool is
+			// complete or not.
+			isIncomplete = poolIsIncomplete(slicesForPool)
+			// If the pool is truly incomplete after this check, then we need to find out if
+			// the slices we have in the pool object really is on the latest generation. If one of
+			// the slices that was filtered out earlier are on a newer generation, it is not.
```

Comment:
We don't get here if we previously picked a stale pool which happens to be complete.

Reply:
I'm not sure I follow exactly how we get into this situation.

If all slices in the pool are targeting the current node, then all the slices will go through `addSlice`, and the existing logic should be correct, i.e. the pool object will contain the slices with the latest generation and the existing logic will correctly identify it as incomplete if any slices are missing (and the second check here should confirm that).

If any slices in the pool doesn't target the current node, then the pool will always be at first identified as incomplete (although perhaps incorrectly). I think this logic should correctly decide if the pool is in fact incomplete or stale and then adjust the slices accordingly (which can lead to pools without any slices).